Initial polynomial-functor representation and specification of Nock

rokopt opened the pull request:
For #1844 .

Introduce:

- A generic type of open terms of any polynomial type, implemented as a functor (type constructor) and its free monad and initial algebra
- A representation of a simple form of non-dependent polynomial functors
- A representation of Nock terms (nouns) as a polynomial
- Conversions between the polynomial and Elixir-native (`Noun.t()`) formats
- Examples/tests for the above

This produces an open "Nock" term -- a Nock syntax with metavariables -- and as such will be used in a follow-on PR to support #1846 .  It also serves as the beginning of a type system including recursive and corecursive types (see #1845 and #1849 , which will also be built upon this in future PRs).

rokopt committed:
Introduce new files

Add skeleton files to which to are to be introduced a
polynomial-functor representation of Nock, and examples of
its use.

rokopt committed:
Introduce generic polynomial term type

A form of S-expression corresponding to types generated
by polynomial functors.  Viewing such types as data types,
it is a term comprising a choice of constructor and
some finite number of fields.

rokopt committed:
Catamorphism for NockPoly term

Also use it for a couple common applications.

rokopt committed:
Initial tests for NockPoly

rokopt committed:
Factor out term-generating functor and its algebra

This separates the functor which generates terms from the
operation of taking its fixed point to produce terms.

rokopt committed:
Factor out the free monad of `termf`

We now generate `t` by applying the free monad to `none()`.
This allows us to define a type `tv` which represents open
terms of `t` -- that is, terms of `t` which may contain
variables drawn from a type parameter.

rokopt committed:
Introduce aliases for open terms

rokopt committed:
`eval` universal morphism for `tv`

(Then reimplement `cata` in terms of `eval`, and
extend `depth` and `size` to operate on open terms.)

rokopt committed:
Add some tests for open terms

rokopt committed:
Add a direct example of `cata`

rokopt committed:
Introduce simplest polynomial-functor spec

This form allows only comparing terms to those generated by
non-dependent, singly-recursive finitary polynomial functors.

rokopt committed:
Introduce initial polynomial Nock-term spec

This represents the type of Nock nouns (`Noun.t()`) in terms of
the polynomial functor of which it is the initial algebra.  (In
particular this means we can also generate its free monad, which
generates its open terms -- Nock terms containing (meta)variables.)

rokopt committed:
Nock term conversions

Conversions between Elixir-native Nock `Noun.t()` and
the polynomial representative thereof (`nock_poly_term`).
The conversion to native expects the caller to have
typechecked the polynomial term as representing a Nock
term, while the conversion from native is obliged to
produce a polynomial term which passes that typecheck.
The two conversions are obliged to be inverses (assuming
they are called on valid inputs).

rokopt committed:
Add some examples of `FinPolyF` usage

rokopt committed:
Introduce tests for polynomial Nock terms

Confirm that we can convert back and forth between the
Elixir-native `Noun.t()` and the polynomial formula for
Nock terms.

rokopt commented:
(Note that the build is currently failing, but it appears to me to be something in the `protobuf` library.)

mariari commented on apps/anoma_lib/lib/examples/enock_poly.ex:
```diff
@@ -2,7 +2,34 @@ defmodule Examples.ENockPoly do
+    assert Term.depth(t4) == 3
+    assert Term.size(t4) == 6
+
+    true
```
I recommend, breaking this into 4 examples...

The main benefit is that you would be able to invoke in your repl the proper term and verify the results yourself!!


```elixir
  def nested_term() do
    res =  {0, [{1, []}, {2, [{3, []}, {4, []}]}, {5, []}]}
    assert Term.depth(res) == 3
    assert Term.size(res) == 6

    res
  end
iex(mariari@Gensokyo)11> Examples.ENockPoly.nested_term
{0, [{1, []}, {2, [{3, []}, {4, []}]}, {5, []}]}
iex(mariari@Gensokyo)12> Examples.ENockPoly.nested_term |> NockPoly.Term.size
6
```

Of course I can then use this to build up more examples

rokopt replied:
Good idea; done in https://github.com/anoma/anoma/pull/1859/commits/d06701a9f9860cce9c1bf86cfcf51a4d9f831f58 .

rokopt committed:
Break up conglomerate tests into smaller ones

In particular this makes it straightforward to capture
values produced by individual examples in the REPL, and
manipulate them from there.

rokopt committed:
Replace third with first person in some comments

rokopt commented on apps/anoma_lib/lib/nock_poly.ex:
```diff
@@ -32,10 +32,25 @@ defmodule NockPoly do
+    type `ctor`, the position-type of that functor is `ctor x Nat`, and
+    for each position `(c, n)`, the direction-type is `Fin n` (finite sets
+    of size `n`).  (In particular it is itself manifestly finitary.)
+    We call this functor `termf` below.
```
Done in 91f71b8e63f3787c5cff7e6bd414df458b0ae00b , along with several other such changes.

rokopt commented on apps/anoma_lib/lib/nock_poly.ex:
```diff
@@ -32,10 +32,25 @@ defmodule NockPoly do
+    We call this functor `termf` below.
     """
 
+    @typedoc "Functor which generates a generic polynomial term parameterized on a constructor type."
```
Also in 91f71b8e63f3787c5cff7e6bd414df458b0ae00b.

rokopt commented on apps/anoma_lib/lib/nock_poly.ex:
```diff
@@ -32,10 +32,25 @@ defmodule NockPoly do
+    @typedoc "Functor which generates a generic polynomial term parameterized on a constructor type."
+    @type termf(ctor, x) :: {ctor, [x]}
+
+    @typedoc "Type of algebras of `termf`."
```
Also in 91f71b8e63f3787c5cff7e6bd414df458b0ae00b.

mariari commented on apps/anoma_lib/lib/examples/enock_poly.ex:
```diff
@@ -11,211 +11,291 @@ defmodule Examples.ENockPoly do
+  t4: a more complex closed term.
+  """
+  def term_test_t4() do
+    res = {0, [{1, []}, {2, [{3, []}, {4, []}]}, {5, []}]}
```
```elixir
res = {0, [term_test_t3(), {2, [{3, []}, {4, []}]}, {5, []}]}
```

we can use the previous example here!

rokopt replied:
I assumed you meant `t1` rather than `t3` and did this in 6d95ab222604901bee7783c531111abe51ea0174.

mariari commented on apps/anoma_lib/lib/examples/enock_poly.ex:
```diff
@@ -11,211 +11,291 @@ defmodule Examples.ENockPoly do
+  t2: a closed term with two children.
+  """
+  def term_test_t2() do
+    res = {:a, [{:a, []}, {:b, []}]}
```
Can we compose this example out of `term_test_t1/0`?

What do we show by doing so, what does it demonstrate about how the API works?

mariari reviewed (commented):
Thanks for your changes, left a few more comments.

My next comment is can we utilize composition in our examples? What do they demonstrate?

rokopt committed:
One illustration of using a term to build others

rokopt commented:
> Thanks for your changes, left a few more comments.
> 
> My next comment is can we utilize composition in our examples? What do they demonstrate?

In many ways!  For example:

- `termf` is a bifunctor, whose `bimap` transforms leaves (constructors) and internal nodes (lists)
- For any `ctor`, `tv(ctor)` is a monad, whose `bind` implements substitution of terms for variables
- For any `ctor`, `ctor'`, `termf(ctor) . termf(ctor')` is also a functor which generates terms, and this functor composition implements _type-level_ substitution
- Any sets of types defined by `tspec`s have:
  - coproducts, which give data types à la carte
  - products, which give trees that grow
  - hom-objects, which give function types
  - parallel products, which compose dynamical systems
 
 ...and that is far from an exhaustive list.  I figured that all these combinators would be best broken up into several PRs and that I would just have to pick a place to call this one done.  My plan was to do the universal morphisms leading up to the `bind` next because that is what provides substitution for macros, use those to do macros themselves, and then gradually add more combinators either as they become necessary for the next features in line or just when it seems like the right time to build them on top of something else that's goign in.  If there are specific ones of these that you think belong in this original PR, please let me know which and I'll extend it.

rokopt commented:
I'm going to close this PR because the branch from which it requests to merge has been superseded by https://github.com/anoma/anoma/tree/terence/inductive-inductive-types .  The latter branch includes all the changes from the one in this PR as well as more enhancements and refactorings and responses to code-review comments.  Once there is a PR from that branch, it will do everything proposed in this PR (and more).
